### PR TITLE
C21 cards

### PR DESCRIPTION
**Pull request by arketec**
Implemented Osgir, The Reconstructor

### CODE REVIEW

File: Mage.Sets/src/mage/sets/Commander2021Edition.java [hunk 1]
```diff
@@ -212,6 +213,7 @@ private Commander2021Edition() {
         cards.add(new SetCardInfo("Orzhov Advokist", 99, Rarity.UNCOMMON, mage.cards.o.OrzhovAdvokist.class));
         cards.add(new SetCardInfo("Orzhov Basilica", 308, Rarity.UNCOMMON, mage.cards.o.OrzhovBasilica.class));
         cards.add(new SetCardInfo("Orzhov Signet", 254, Rarity.UNCOMMON, mage.cards.o.OrzhovSignet.class));
+        cards.add(new SetCardInfo("Osgir, the Reconstructor", 8, Rarity.MYTHIC, OsgirTheReconstructor.class));
```
Comment (theelk801) [inline Mage.Sets/src/mage/sets/Commander2021Edition.java:216]: replace `OsgirTheReconstructor.class` with `mage.cards.o.OsgirTheReconstructor.class` to eliminate the import

File: Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    public OsgirTheReconstructorCreateArtifactTokensEffect(final OsgirTheReconstructorCreateArtifactTokensEffect effect)  {
+        super(effect);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        Card card = game.getCard(targetPointer.getFirst(game, source));
+
+        player.moveCards(card, Zone.EXILED, source, game);
```
Comment (theelk801) [inline Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:114]: The card needs to be exiled as part of the cost, not the effect. You need to created a `CostAdjuster` to go with the `TargetAdjuster`. Take a look at cards like Necropolis Fiend and Scorched Earth.

File: Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        this.power = new MageInt(4);
+        this.toughness = new MageInt(4);
+
+        // Vigilance
+        this.addAbility(VigilanceAbility.getInstance());
+
+        // {1}, Sacrifice an artifact: Target creature you control gets +2/+0 until end of turn.
+        Ability ability = new SimpleActivatedAbility(Zone.BATTLEFIELD, new BoostTargetEffect(2, 0, Duration.EndOfTurn), new ManaCostsImpl<>("{1}"));
+        ability.addTarget(new TargetCreaturePermanent());
+        ability.addCost(new SacrificeTargetCost(new TargetControlledPermanent(new FilterControlledArtifactPermanent("an artifact"))));
```
Comment (theelk801) [inline Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:62]: You can use `StaticFilters.FILTER_CONTROLLED_PERMANENT_ARTIFACT_AN` here

File: Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+        CreateTokenCopyTargetEffect effect = new CreateTokenCopyTargetEffect(player.getId(), null, false, 2);
+        effect.setTargetPointer(new FixedTarget(card.getId(), game.getState().getZoneChangeCounter(card.getId())));
+        effect.apply(game, source);
+
+        return  true;
+    }
+
+    @Override
+    public Effect copy() {
```
Comment (theelk801) [inline Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:128]: this should return `OsgirTheReconstructorCreateArtifactTokensEffect`

**Review by theelk801 — CHANGES_REQUESTED** [review]

File: Mage.Sets/src/mage/sets/Commander2021Edition.java (hunk 1 above)
Comment (arketec) [reply Mage.Sets/src/mage/sets/Commander2021Edition.java:216]: no problem, this has been updated

File: Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java (hunk 3 above)
Comment (arketec) [reply Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:62]: this has also been updated

File: Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java (hunk 4 above)
Comment (arketec) [reply Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:128]: this does return that no?

File: Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java (hunk 2 above)
Comment (arketec) [reply Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:114]: everytime I try to add the exiled card as a cost, I cannot seem to be able to get it to copy. changed adjustTargets to this:

public void adjustTargets(Ability ability, Game game) {
        int xValue = ability.getManaCostsToPay().getX();
        FilterCard filter = new FilterArtifactCard("an artifact card with mana value "+xValue+" from your graveyard");
        filter.add(new ManaValuePredicate(ComparisonType.EQUAL_TO, xValue));

        ability.getTargets().clear();
        for (Cost cost : ability.getCosts()) {
            if (!(cost instanceof ExileFromGraveCost)) {
                continue;
            }
            ExileFromGraveCost exileCost = (ExileFromGraveCost) cost;
            for (Target target : exileCost.getTargets()) {
                if (target instanceof TargetCardInYourGraveyard) {
                    ability.getTargets().add(target);
                }
            }
        }
    }

but it does not target anything

also, is there a way to set breakpoints inside Mage.Sets? The debugger seems to skip righ over those

File: Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java (hunk 4 above)
Comment (theelk801) [reply Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:128]: I mean that it should be `public OsgirTheReconstructorCreateArtifactTokensEffect copy() {`, but I'm not sure if it even matters

Comment (arketec) [reply Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:128]: got it

File: Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java (hunk 2 above)
Comment (theelk801) [reply Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:114]: Oh I just realized that you don't need a `TargetAdjuster` at all, you just need a `CostAdjuster`. You might actually want to create a custom cost instead like Marath, Will of the Wild has. 

As for breakpoints, I'm a shitty dev who uses print statements so I can't answer that.

Comment (arketec) [reply Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java:114]: haha, alright thanks

Comment (arketec) [reply Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java]: ok, I ended up being able fix it with just a cost adjuster. submitted all changes

**Pull request merged**